File: shortest_dist_2.py
# Time Complexity :
# O(N)

# Space Complexity :
# O(1)

# Did this code successfully run on Leetcode :
#Yes

#We go through each word and store it's location
#when the request for shortest distance between 2 words comes in, we go through the list of locations of these 2 words using 2 pointers. The pointer of the list whose value is lower is increased and the minimum distance is updated at each step
#In the end, we return the min distance computed

class WordDistance:

    def __init__(self, wordsDict: List[str]):
        self.loc_map = {}
        self.max_dist = len(wordsDict)
        for i,word in enumerate(wordsDict) :
            if word not in self.loc_map :
                self.loc_map[word] = []
            self.loc_map[word].append(i)
        
    def shortest(self, word1: str, word2: str) -> int:
        locs_1 = self.loc_map[word1]
        locs_2 = self.loc_map[word2]
        min_dist = self.max_dist
        p1 = 0
        p2 = 0
        while (p1 < len(locs_1) and p2 < len(locs_2)):
            curr_dist = abs(locs_1[p1] - locs_2[p2] )
            min_dist = min(min_dist, curr_dist )
            if locs_1[p1] < locs_2[p2] :
                p1 += 1
            else :
                p2 += 1
        return min_dist

# Your WordDistance object will be instantiated and called as such:
# obj = WordDistance(wordsDict)
# param_1 = obj.shortest(word1,word2)

File: shortestDist2.java
//Time: O(n)
//Space: O(1)
//Program ran on leetcode successfully

class shortestDist2 {
    private HashMap<String, ArrayList<Integer>> map;

    public WordDistance(String[] words) {
        map = new HashMap<>();
        for (int i = 0; i < words.length; i++) {
            if (!map.containsKey(words[i])) {
                map.put(words[i], new ArrayList<Integer>());
            }
            map.get(words[i]).add(i);
        }
    }

    public int shortest(String word1, String word2) {
        List<Integer> p1 = map.get(word1);
        List<Integer> p2 = map.get(word2);
        if (p1 == null || p2 == null) {
            return -1;
        }
        int shortest = Integer.MAX_VALUE;
        int i1 = 0;
        int i2 = 0;
        while (i1 < p1.size() && i2 < p2.size()) {
            shortest = Math.min(Math.abs(p1.get(i1) - p2.get(i2)), shortest);
            if (p1.get(i1) < p2.get(i2)) {
                i1++;
            } else {
                i2++;
            }
        }
        return shortest;
    }
}